### PR TITLE
Refactor: extend error validation of query parameters

### PR DESCRIPTION
**Pull request by delveper**
Refactoring according to recommendations given during the last demo.

### CODE REVIEW

**Review by Brigant — APPROVED** [review]

File: app/entities/query.go [hunk 1]
```diff
@@ -16,22 +16,30 @@ type QueryParams struct {
...
+	if qp.Offset == "" {
+		return fmt.Errorf("offset argument has to be set")
+	}
+
 	if _, err := strconv.Atoi(qp.Offset); err != nil {
 		return fmt.Errorf("offset argument has to be a number")
 	}
 
-	allowedSortArgs := []string{"first_name", "last_name", "created_at", ","}
+	allowedSortArgs := []string{",", "first_name", "last_name", "created_at"}
```
Comment (andstepko) [inline app/entities/query.go:35]: Coma is not a sort argument.

File: app/entities/query.go [hunk 2]
```diff
@@ -16,22 +16,30 @@ type QueryParams struct {
...
 
 	for _, arg := range allowedSortArgs {
 		qp.Sort = strings.ReplaceAll(qp.Sort, arg, "")
 	}
 
 	if len(qp.Sort) > 0 {
-		return fmt.Errorf("sort argument `%v` does not fit list: %v", qp.Sort, allowedSortArgs)
+		return fmt.Errorf("sort argument `%v` does not fit list: %+v", qp.Sort, allowedSortArgs[1:])
```
Comment (andstepko) [inline app/entities/query.go:42]: That's a dirty trick 😄 
Just don't have coma in the `allowedSortArgs`. It's a delimiter not a value.

File: app/entities/query.go [hunk 3]
```diff
@@ -16,22 +16,28 @@ type QueryParams struct {
...
 	if _, err := strconv.Atoi(qp.Offset); err != nil {
 		return fmt.Errorf("offset argument has to be a number")
 	}
 
-	allowedSortArgs := []string{"first_name", "last_name", "created_at", ","}
-
-	for _, arg := range allowedSortArgs {
-		qp.Sort = strings.ReplaceAll(qp.Sort, arg, "")
-	}
+	allowedSortArgs := map[string]bool{"first_name": true, "last_name": true, "created_at": true}
```
Comment (andstepko) [inline app/entities/query.go:35]: If it is a Set (data structure with values that don't duplicate) - use `map[string]struct{}` (empty struct a) weights 0 bytes b) obvious that there won't be both true and false entries there).

Comment (delveper) [reply app/entities/query.go:35]: I have already tried usign {} and it looks ugly when I print it out.

Comment (andstepko) [reply app/entities/query.go:35]: I often use slices for that things, to be honest.

Comment (delveper) [reply app/entities/query.go:35]: O^2 complexity is not acceptable 😄

**Review by andstepko — APPROVED** [review]

File: app/entities/query.go [hunk 4]
```diff
@@ -16,22 +16,32 @@ type QueryParams struct {
...
+	}
+
 	if _, err := strconv.Atoi(qp.Offset); err != nil {
 		return fmt.Errorf("offset argument has to be a number")
 	}
 
-	allowedSortArgs := []string{"first_name", "last_name", "created_at", ","}
+	allowedSortArgs := []string{"first_name", "last_name", "created_at"}
+
+	const sep = ","
```
Comment (andstepko) [inline app/entities/query.go:37]: September?

Comment (delveper) [reply app/entities/query.go:37]: [sep](https://www.abbreviations.com/abbreviation/separator) Nice try 😃

Comment (delveper) [reply app/entities/query.go:37]: btw it is not only my imagination. 
according to *stdlib*:
```go
func Join(elems []string, sep string) string
```

Comment (Rusli4k) [reply app/entities/query.go:37]: maybe "separ" ? )))

Comment (delveper) [reply app/entities/query.go:37]: Wait, how'd you know? 😄

Comment (vpakhuchyi) [reply app/entities/query.go:37]: Constant name can be longer than func argument name.

Comment (vpakhuchyi) [reply app/entities/query.go:37]: Constant names always should be descriptive. Func arguments names may be less descriptive, cos func documentation can contains some additional information and make understanding to be easier.

Comment (delveper) [reply app/entities/query.go:37]: It is a huge point. thanks

**Pull request closed without merging**